File: 프로그래머스/3/72415. 카드 짝 맞추기/카드 짝 맞추기.kt
import kotlin.math.abs
import kotlin.math.min

class Solution {
    private val dr = intArrayOf(-1, 1, 0, 0)
    private val dc = intArrayOf(0, 0, -1, 1)

    private fun idx(r: Int, c: Int) = r * 4 + c
    private fun inRange(r: Int, c: Int) = r in 0..3 && c in 0..3

    private fun ctrlMove(from: Int, dir: Int, mask: Int): Int {
        var r = from / 4
        var c = from % 4
        while (true) {
            val nr = r + dr[dir]
            val nc = c + dc[dir]
            if (!inRange(nr, nc)) break
            r = nr; c = nc
            val ni = idx(r, c)
            if ((mask and (1 shl ni)) != 0) return ni // 카드 발견
        }
        return idx(r, c) // 끝 칸
    }

    private fun stepMove(from: Int, dir: Int): Int {
        val r = from / 4
        val c = from % 4
        val nr = r + dr[dir]
        val nc = c + dc[dir]
        if (!inRange(nr, nc)) return from
        return idx(nr, nc)
    }

    private fun dist(start: Int, target: Int, mask: Int): Int {
        if (start == target) return 0
        val vis = IntArray(16) { -1 }
        val q = IntArray(16)
        var head = 0
        var tail = 0
        q[tail++] = start
        vis[start] = 0

        while (head < tail) {
            val u = q[head++]
            val d = vis[u]
            for (dir in 0..3) {
                val v1 = stepMove(u, dir)
                if (vis[v1] == -1) {
                    vis[v1] = d + 1
                    if (v1 == target) return vis[v1]
                    q[tail++] = v1
                }
                val v2 = ctrlMove(u, dir, mask)
                if (vis[v2] == -1) {
                    vis[v2] = d + 1
                    if (v2 == target) return vis[v2]
                    q[tail++] = v2
                }
            }
        }
        return vis[target]
    }

    private fun key(mask: Int, cursor: Int): Long = (mask.toLong() shl 4) or cursor.toLong()

    private lateinit var types: IntArray
    private lateinit var pos: Array<IntArray> // pos[t] = intArrayOf(p1,p2)
    private val memo = HashMap<Long, Int>()

    private fun dfs(mask: Int, cursor: Int): Int {
        if (mask == 0) return 0
        val k = key(mask, cursor)
        memo[k]?.let { return it }

        var best = Int.MAX_VALUE

        for (t in types) {
            val p1 = pos[t][0]
            val p2 = pos[t][1]
            val has1 = (mask and (1 shl p1)) != 0
            val has2 = (mask and (1 shl p2)) != 0
            if (!has1 && !has2) continue

            val nextMask = mask and (1 shl p1).inv() and (1 shl p2).inv()

            // 순서 1: cursor -> p1 (Enter) -> p2 (Enter)
            val cost1 = dist(cursor, p1, mask) + 1 + dist(p1, p2, mask) + 1
            best = min(best, cost1 + dfs(nextMask, p2))

            // 순서 2: cursor -> p2 (Enter) -> p1 (Enter)
            val cost2 = dist(cursor, p2, mask) + 1 + dist(p2, p1, mask) + 1
            best = min(best, cost2 + dfs(nextMask, p1))
        }

        memo[k] = best
        return best
    }

    fun solution(board: Array<IntArray>, r: Int, c: Int): Int {
        // 카드 위치 수집 + 마스크 구성
        val tmpPos = Array(7) { ArrayList<Int>(2) }
        var mask = 0
        val exist = BooleanArray(7)

        for (i in 0 until 4) {
            for (j in 0 until 4) {
                val v = board[i][j]
                if (v != 0) {
                    val p = idx(i, j)
                    mask = mask or (1 shl p)
                    exist[v] = true
                    tmpPos[v].add(p)
                }
            }
        }

        // types, pos 준비
        val typeList = ArrayList<Int>()
        for (v in 1..6) if (exist[v]) typeList.add(v)
        types = typeList.toIntArray()

        pos = Array(7) { intArrayOf(0, 0) }
        for (v in 1..6) {
            if (tmpPos[v].size == 2) {
                pos[v] = intArrayOf(tmpPos[v][0], tmpPos[v][1])
            }
        }

        memo.clear()
        return dfs(mask, idx(r, c))
    }
}

File: 프로그래머스/3/72415. 카드 짝 맞추기/카드 짝 맞추기.js
function solution(board, r, c) {
  const N = 4;
  const dr = [-1, 1, 0, 0];
  const dc = [0, 0, -1, 1];
  const idx = (rr, cc) => rr * 4 + cc;
  const inRange = (rr, cc) => rr >= 0 && rr < 4 && cc >= 0 && cc < 4;

  // 남은 카드 마스크 + 각 숫자(1~6)의 위치 2개 저장
  let mask = 0;
  const pos = Array.from({ length: 7 }, () => []);
  for (let i = 0; i < 4; i++) {
    for (let j = 0; j < 4; j++) {
      const v = board[i][j];
      if (v !== 0) {
        const p = idx(i, j);
        mask |= (1 << p);
        pos[v].push(p);
      }
    }
  }

  const types = [];
  for (let v = 1; v <= 6; v++) if (pos[v].length === 2) types.push(v);

  function ctrlMove(from, dir, curMask) {
    let rr = Math.floor(from / 4);
    let cc = from % 4;
    while (true) {
      const nr = rr + dr[dir];
      const nc = cc + dc[dir];
      if (!inRange(nr, nc)) break;
      rr = nr; cc = nc;
      const ni = idx(rr, cc);
      if (curMask & (1 << ni)) return ni; // 카드 발견
    }
    return idx(rr, cc); // 끝칸
  }

  function stepMove(from, dir) {
    const rr = Math.floor(from / 4);
    const cc = from % 4;
    const nr = rr + dr[dir];
    const nc = cc + dc[dir];
    if (!inRange(nr, nc)) return from;
    return idx(nr, nc);
  }

  // BFS: start -> target 최소 이동 횟수
  function dist(start, target, curMask) {
    if (start === target) return 0;
    const vis = Array(16).fill(-1);
    const q = Array(16);
    let head = 0, tail = 0;

    q[tail++] = start;
    vis[start] = 0;

    while (head < tail) {
      const u = q[head++];
      const d = vis[u];

      for (let dir = 0; dir < 4; dir++) {
        const v1 = stepMove(u, dir);
        if (vis[v1] === -1) {
          vis[v1] = d + 1;
          if (v1 === target) return vis[v1];
          q[tail++] = v1;
        }
        const v2 = ctrlMove(u, dir, curMask);
        if (vis[v2] === -1) {
          vis[v2] = d + 1;
          if (v2 === target) return vis[v2];
          q[tail++] = v2;
        }
      }
    }
    return vis[target];
  }

  // 메모: key = (mask<<4) | cursor
  const memo = new Map();
  const startCursor = idx(r, c);

  function dfs(curMask, cursor) {
    if (curMask === 0) return 0;
    const key = (curMask << 4) | cursor;
    if (memo.has(key)) return memo.get(key);

    let best = Infinity;

    for (const t of types) {
      const [p1, p2] = pos[t];
      const has1 = (curMask & (1 << p1)) !== 0;
      const has2 = (curMask & (1 << p2)) !== 0;
      if (!has1 && !has2) continue;

      const nextMask = curMask & ~(1 << p1) & ~(1 << p2);

      // 순서 1: cursor -> p1 (Enter) -> p2 (Enter)
      const cost1 = dist(cursor, p1, curMask) + 1 + dist(p1, p2, curMask) + 1;
      best = Math.min(best, cost1 + dfs(nextMask, p2));

      // 순서 2: cursor -> p2 (Enter) -> p1 (Enter)
      const cost2 = dist(cursor, p2, curMask) + 1 + dist(p2, p1, curMask) + 1;
      best = Math.min(best, cost2 + dfs(nextMask, p1));
    }

    memo.set(key, best);
    return best;
  }

  return dfs(mask, startCursor);
}